includes/class-wcs-object-data-cache-manager.php

```diff
@@ -0,0 +1,208 @@
+<?php
+if ( ! defined( 'ABSPATH' ) ) {
+	exit;
+}
+
+/**
+ * Class for managing caches of object data.
+ *
+ * This class will track changes to an object (specified by the object type value) and trigger an action hook for each change to any specific meta key or object (specified by the $data_keys variable).
+ * Interested parties (like our cache store classes), can then listen for these hooks and update their caches accordingly.
+ *
+ * @version  5.2.0
+ * @category Class
+ */
+class WCS_Object_Data_Cache_Manager extends WCS_Post_Meta_Cache_Manager {
+
+	/**
+	 * The WC_Data object type this cache manager will track changes to. eg 'order', 'subscription'.
+	 *
+	 * @var string
+	 */
+	protected $object_type;
+
+	/**
+	 * The object's data keys this cache manager will keep track of changes to. Can be an object property key ('customer_id') or meta key ('_subscription_renewal').
+	 *
+	 * @var array
+	 */
+	protected $data_keys;
+
+	/**
+	 * An internal record of changes to the object that this manager is tracking.
+	 *
+	 * This internal record is generated before the object is saved, so we can determine
+	 * if the value has changed, what the previous value was, and what the new value is.
+	 *
+	 * In the event that the object is being created (doesn't have an ID prior to save), this
+	 * record will be generated after the object is saved, and all the data this manager
+	 * is tracking will be pulled from the created object.
+	 *
+	 * @var array Each element is keyed by the object's ID, and contains an array of tracked changes {
+	 *     Data about the change that was made to the object.
+	 *
+	 *     @type mixed  $new      The new value.
+	 *     @type mixed  $previous The previous value before it was changed.
+	 *     @type string $type     The type of change. Can be 'update', 'add' or 'delete'.
+	 * }
+	 */
+	protected $object_changes = [];
+
+	/**
+	 * Constructor.
+	 *
+	 * @param string The post type this cache manage acts on.
+	 * @param array The post meta keys this cache manager should act on.
+	 */
+	public function __construct( $object_type, $data_keys ) {
+		$this->object_type = $object_type;
+		$this->data_keys   = $data_keys;
+	}
+
+	/**
+	 * Attaches callbacks to keep the caches up-to-date.
+	 */
+	public function init() {
+		add_action( "woocommerce_before_{$this->object_type}_object_save", [ $this, 'prepare_object_changes' ] );
+		add_action( "woocommerce_after_{$this->object_type}_object_save", [ $this, 'action_object_cache_changes' ] );
```

Comment:
I haven't tested this yet but I just wanted to put this question down in GH first before I forget to come back to it, but do these hooks also get triggered when `$subscription->save_meta_data()` is called?

i.e.

```
$renewal_order->update_meta_data( '_subscription_renewal', 123 );
$renewal_order->save_meta_data();
```

Reply:
Based on the nature of this PR having most of the code/functionality behind the HPOS feature flag, I'm wondering if we should treat this PR like a solid base and then open issues for these things separately if they exist?

If that's the case I'm happy to merge this PR as is :)

Reply:
> Based on the nature of this PR having most of the code/functionality behind the HPOS feature flag, I'm wondering if we should treat this PR like a solid base and then open issues for these things separately if they exist?

Absolutely. There are lots of little things still going on with this PR, like CPT references in class names, post meta mentioned in functions etc. 

I've described working on this PR like pulling on a piece of string, we need to stop somewhere and merging this as a base, safely given it's locked behind HPOS is good with me.